File: example.java/src/main/java/ru/model/GenericModel.java
package ru.model;

import lombok.*;
import java.util.*;

@Value
@Builder
public class GenericModel<F, S extends GenericTypeParameter<F>> {
    F first;
    S second;

    F[] fistArray;
    S[] secondArray;

    List<F> firstList;
    List<S> secondList;

    GenericModel<F, S> firstModel;
    GenericModel<F, S>[] firstModelArray;

    List<GenericModel<F, S>>[] firstModelList;
    Set<GenericModel<F, S>>[] firstModelSet;
    Collection<GenericModel<F, S>>[] firstModelCollection;
    Map<String, GenericModel<F, S>>[] firstModelMap;
}

File: example.kotlin/src/main/java/ru/model/GenericModel.kt
package ru.model

data class GenericModel<F, S : GenericTypeParameter<F>>(
        val first: F,
        val second: S,
        val fistArray: Array<F>,
        val secondArray: Array<S>,
        val firstList: List<F>,
        val secondList: List<S>,
        val firstModel: GenericModel<F, S>,
        val firstModelArray: Array<GenericModel<F, S>>,
        val firstModelList: Array<List<GenericModel<F, S>>>,
        val firstModelSet: Array<Set<GenericModel<F, S>>>,
        val firstModelCollection: Array<Collection<GenericModel<F, S>>>,
        val firstModelMap: Array<Map<String, GenericModel<F, S>>>,
)
